### PR TITLE
Cavity flow PINNs example

### PR DESCRIPTION
**Pull request by conordaly0**

### CODE REVIEW

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+ $$ \frac{\partial u}{\partial x}+\frac{\partial v}{\partial y}=0 $$ 
+
+ $$ u\frac{\partial u}{\partial x}+v\frac{\partial u}{\partial y}+\frac{\partial p}{\partial x}-\frac{1}{Re}\bigg(\frac{\partial^2 u}{\partial x^2 }+\frac{\partial^2 u}{\partial y^2 }\bigg)=0 $$ 
+
+ $$ u\frac{\partial v}{\partial x}+v\frac{\partial v}{\partial y}+\frac{\partial p}{\partial y}-\frac{1}{Re}\bigg(\frac{\partial^2 v}{\partial x^2 }+\frac{\partial^2 v}{\partial y^2 }\bigg)=0 $$ 
+
+`(x,y)` are the spatial coordinates, `(u,v)` is the fluid velocity, `p` is the pressure, `Re` is the Reynolds number.
+
+
+In order to automatically satisfy the continuity equation we use the stream function psi such that `u=psi_y` and `v=-psi_x`. The boundary conditions are `(u,v)=(1,0)` and the top boundary and `(u,v)=(0,0)` at the other boundaries. The Reynolds number `Re=100`.
```
Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:18]: Typo: "and the top boundary" should be "at the top boundary"

Comment (mmarkows17) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:18]: Similarly, put an "is" between Reynolds number and Re=100

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+Re = 100;
+u0 = 1;
+```
+# Create network
+```matlab
+% Create basic MLP network architecture with two inputs (x,y) and two
+% outputs (psi,p).
+numHiddenUnits = 32;
+net = dlnetwork();
+layers = [ featureInputLayer(1, Name="x")
```
Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:34]: I'm a little confused by what is going on with the architecture, it would be helpful if the architecture setup had more description. For instance, why not just put a featureInputLayer(2) as the first layer?

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    swishLayer(Name="swishout")
+    fullyConnectedLayer(1, Name="psiFree") ];
+net = addLayers(net, layers);
+net = addLayers(net, fullyConnectedLayer(1, Name="p"));
+net = connectLayers(net, "swishout", "p");
+net = addInputLayer(net, featureInputLayer(1, Name="y"), Initialize=false);
+
+% Add anchor functions to strictly enforce boundary conditions on the
+% streamfunction.
+net = addLayers(net, [functionLayer(@(x)4.*x.*(1-x), Name="anchorX", Acceleratable=true); multiplicationLayer(3, Name="psi")]);
```
Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:52]: Similarly here, I think this needs some more explanation. This function is chosen to satisfy the boundary conditions? Can you give a short description about what the multiplicationLayer and connectLayers do?

Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:52]: I would also highlight that this can be viewed as a hard enforcement of the physics, i.e. some physics get enforced in the architecture. I think this is the only published PINN example where we do this, the rest are soft enforcement via the loss function.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+function [loss, grads, lossEqnX, lossEqnY, lossBC] = pinnsLossFunction(net, xyEquation, xyBoundary, zeroVector, uvBoundary, Re)
+
+% Get model outputs at interior points.
+xeq = xyEquation(:, 1);
+yeq = xyEquation(:, 2);
+[psi, p] = forward(net, xeq, yeq);
+
+% Compute gradients.
+psisum = sum(psi,1);
+u = dlgradient(psisum, yeq, EnableHigherDerivatives=true);
```
Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:211]: Why not use dljacobian?

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 5]
```diff
@@ -0,0 +1,274 @@
+
+# Cavity flow with Physics\-Informed Neural Networks
+
+Solve cavity flow governed by 2d steady Navier\-Stokes equations and continuity equation, using a Physics\-Informed Neural Network (PINN).
```
Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:4]: Are there any Github repos or papers that should be cited in the example?

File: physics-informed-neural-networks-for-steady-cavity-flow/images/figure_0.png
Comment (mmarkows17) [inline physics-informed-neural-networks-for-steady-cavity-flow/images/figure_0.png:1]: Nice pictures. How good does the PINN do? Any baseline to compare it to?

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md (hunk 1 above)
Comment (bwdGitHub) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:18]: It's nice to use LaTeX for $\psi$ and $u = \psi_y$ etc. or monospace `psi` if referring to a variable.

**Review by mmarkows17 — COMMENTED** [review]
Hi Conor, thank you for putting this together!

My biggest suggestion is to add more description/explanation of what is happening in the architecture setup. The rest are smaller comments for your consideration.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+net = connectLayers(net, "y", "anchorY");
+net = connectLayers(net, "anchorY", "psi/in2");
+net = connectLayers(net, "psiFree", "psi/in3");
+
+% Make sure outputs are ordered (psi,p).
+net.OutputNames = ["psi", "p"];
+
+% Initialize the network and cast to double precision.
+net = initialize(net);
+net = dlupdate(@double, net);
```
Comment (bwdGitHub) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:64]: Can we add a plot of the network to the readme? I think it'd help make it clearer how the anchors enforce the boundary conditions.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+title('u')
+
+subplot(2,2,4)
+contourf(xt, yt, vt)
+colorbar
+axis equal
+title('v')
+```
+
+![figure_0.png](README_media/figure_0.png)
```
Comment (bwdGitHub) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:199]: This link is broken, should be something like `./images/figure_0.png` for the path.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+loss = mean( log(cosh(e)), 'all' );
+end
+
+function [psi, p, u, v] = calculateStreamfunctionPressureAndVelocity(net, x, y)
+% Compute the streamfunction psi, pressure p and velocity (u,v) given
+% input positions (x,y).
+[psi, p] = forward(net, x, y);
+psisum = sum(psi,1);
+u = dlgradient(psisum, y);
+v = -1.*dlgradient(psisum, x);
```
Comment (bwdGitHub) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:267]: It could be worth making this one `dlgradient` call, something like
```
[u,v] = dlgradient(psisum,y,x);
v = -v;
```
I had some feeling this was faster, although I'm not completely confident. I'm sure I found it faster when I did something like batching the interior points and boundary points together and performing one `dlgradient`, then splitting the output back into interior and boundary gradients. There could be a few places to try that out in the loss above.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md (hunk 4 above)
Comment (bwdGitHub) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:211]: I suppose it would mean you'd need at least R2024b. Otherwise I expect it should be ok - it'll do the `sum(psi,1)` line and the `EnableHigherDerivatives=true` by default, at the expense of the extra dimension argument `dljacobian(psi,yeq,2)`.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md (hunk 1 above)
Comment (conordaly0) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:18]: I think using LaTeX in the text is a nice way to go -- I'll do that.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md (hunk 3 above)
Comment (conordaly0) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:52]: Good point!

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md (hunk 4 above)
Comment (conordaly0) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:211]: I do think `dljacobian` is cleaner here -- I'll switch to that.

File: physics-informed-neural-networks-for-steady-cavity-flow/images/figure_0.png
Comment (conordaly0) [reply physics-informed-neural-networks-for-steady-cavity-flow/images/figure_0.png:1]: I didn't do any quantitative comparison against an FEM baseline. Qualitatively the results look reasonably close to solutions via other methods. While I think it's worth adding some statement to the benefit of the PINNs approach, I'll hold off on making a more detailed comparison against other baselines -- I think that's worthy of it's own submission.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md (hunk 6 above)
Comment (conordaly0) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:64]: Good idea -- I'll add that.

File: physics-informed-neural-networks-for-steady-cavity-flow/README.md [hunk 9]
```diff
@@ -207,25 +199,21 @@ yeq = xyEquation(:, 2);
...
-vsum = sum(v,1);
-vx = dlgradient(vsum, xeq, EnableHigherDerivatives=true);
-vy = dlgradient(vsum, yeq, EnableHigherDerivatives=true);
-vxx = dlgradient(sum(vx,1), xeq, EnableHigherDerivatives=true);
-vyy = dlgradient(sum(vy,1), yeq, EnableHigherDerivatives=true);
-
-psum = sum(p,1);
-px = dlgradient(psum, xeq, EnableHigherDerivatives=true);
-py = dlgradient(psum, yeq, EnableHigherDerivatives=true);
+u = dljacobian(psi', yeq, 1);
```
Comment (bwdGitHub) [inline physics-informed-neural-networks-for-steady-cavity-flow/README.md:202]: If `psi` and `yeq` are both "batch x features" I think `u = dljacobian(psi, yeq, 2)` would work here. But then `u` is "hidden x batch x features" where `hidden==size(psi,2)`, and since you end up wanting more derivatives along that "hidden" dimension, I guess you have to permute or swap to `dljacobian(u,xeq,1)` anyway. I think having stuff as "features x batch" works out more nicely for me typically, which means replacing `featureInputLayer` with `inputLayer` if you want to use unlabelled `dlarray`.

Comment (conordaly0) [reply physics-informed-neural-networks-for-steady-cavity-flow/README.md:202]: I can see how working in "features x batch" is most convenient for the network. On the other hand, the "batch x features" layout is closer to the original data layout. 

I'll keep things as they are for now and can switch to "features x batch" later if we think it is neater.
